extensions/wikia/Rail/RailController.class.php

```diff
@@ -8,139 +8,163 @@ class RailController extends WikiaController {
-				$this->railLazyContent .= $this->app->renderView(
-					$railModule[0], /* Controller */
-					$railModule[1], /* Method */
-					$railModule[2] /* array of params */
-				);
-			}
+			'wgTitle' => $title
+		] );
 
-			$this->railLazyContent .= Html::element('div', ['id' => 'WikiaAdInContentPlaceHolder']);
+		$assetManager = AssetsManager::getInstance();
 
-			$this->css = $sassFiles = [];
-			foreach (array_keys($this->app->wg->Out->styles) as $style) {
-				if ($wgAllInOne && $assetManager->isSassUrl($style)) {
-					$sassFiles[] = $style;
-				} else {
-					$this->css[] = $style;
-				}
-			}
+		$railModuleListRaw = $wrapper->wrap( function () use ( $context ) {
+			return ( new BodyController )->getRailModuleList();
+		} );
 
-			if (!empty($sassFiles)) {
-				$excludeScss = (array) $this->getRequest()->getVal('excludeScss', []);
-				$sassFilePath = (array) $assetManager->getSassFilePath($sassFiles);
-				$includeScss = array_diff($sassFilePath, $excludeScss);
+		$railModules = $this->filterModules(
+			$railModuleListRaw,
+			self::FILTER_LAZY_MODULES
+		);
+		$this->railLazyContent = '';
 
-				// SUS-771: Log any duplicate CSS that rail modules try to load but are already loaded by Oasis skin
-				$duplicateScss = array_intersect( $sassFilePath, $excludeScss );
-				if ( count( $duplicateScss ) ) {
-					Wikia\Logger\WikiaLogger::instance()->info( 'SUS-771', [
-						'styles' => json_encode( $duplicateScss )
-					] );
-				}
+		krsort( $railModules );
 
-				if (!empty($includeScss)) {
-					$this->css[] = $assetManager->getSassesUrl($includeScss);
-				}
-			}
+		// TODO XW-2760 remove after experiment is done
+		$isAdMixExperimentEnabled = $this->request->getVal( 'isAdMixExperimentEnabled', false );
 
-			// Do not load user and site jses as they are already loaded and can break page
-			$oldWgUseSiteJs = $wgUseSiteJs;
-			$oldWgAllowUserJs = $wgAllowUserJs;
-			$wgUseSiteJs = false;
-			$wgAllowUserJs = false;
+		if ( $isAdMixExperimentEnabled ) {
+			// ad slot NATIVE_TABOOLA_RAIL
+			unset( $railModules[1435] );
+			// ad slot LEFT_SKYSCRAPER_2
+			unset( $railModules[1100] );
+
+			// copied from RecirculationHooks::onGetRailModuleList
+			$recirculationModulePosition = $context->getUser()->isAnon() ? 1305 : 1285;
+			unset( $railModules[$recirculationModulePosition] );
+
+			array_push( $railModules, [ 'AdMixExperiment', 'recirculationAndAdPlaceholder' ] );
+		}
+
+		foreach ( $railModules as $railModule ) {
+			$this->railLazyContent .= $this->app->renderView(
+				$railModule[0], /* Controller */
+				$railModule[1], /* Method */
+				$railModule[2] /* array of params */
+			);
+		}
+
+		// ad mix experiment uses a wrapper to group recirculation and ad placeholder
+		if ( !$isAdMixExperimentEnabled ) {
+			$this->railLazyContent .= Html::element( 'div', [ 'id' => 'WikiaAdInContentPlaceHolder' ] );
+		}
+
+		$this->css = $sassFiles = [];
+		foreach ( array_keys( $this->app->wg->Out->styles ) as $style ) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this code to rely on RequestContext instead of context-dependent globals. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.wikia-inc.com/coding_rules#rule_key=wikia-php-rules%3AContextualWgVariableUsage)